zebra-consensus/src/block/tests.rs

```diff
@@ -140,6 +142,109 @@ fn time_check_past_block() {
     // will have a clock that's far enough in the past for the test to
     // fail.
-    check::is_time_valid_at(&block.header, now)
+    check::time_is_valid_at(&block.header, now)
         .expect("the header time from a mainnet block should be valid");
 }
+
+#[test]
+fn subsidy_is_correct_test() -> Result<(), Report> {
+    subsidy_is_correct_for_network(Network::Mainnet)?;
+    subsidy_is_correct_for_network(Network::Testnet)?;
+
+    Ok(())
+}
+
+fn subsidy_is_correct_for_network(network: Network) -> Result<(), Report> {
+    let block_iter = match network {
+        Network::Mainnet => zebra_test::vectors::MAINNET_BLOCKS.iter(),
+        Network::Testnet => zebra_test::vectors::TESTNET_BLOCKS.iter(),
+    };
+    for (&height, block) in block_iter {
+        let block = block
+            .zcash_deserialize_into::<Block>()
+            .expect("block is structurally valid");
+
+        // TODO: first halving, second halving, third halving, and very large halvings
+        if block::Height(height) > SLOW_START_INTERVAL
+            && block::Height(height) < NetworkUpgrade::Canopy.activation_height(network).unwrap()
+        {
+            check::subsidy_is_correct(network, &block)
+                .expect("subsidies should pass for this block");
+        }
+    }
+    Ok(())
+}
+
+#[test]
+fn nocoinbase_validation_failure() -> Result<(), Report> {
+    use crate::error::*;
+
+    let network = Network::Mainnet;
+
+    // Get a header form a block in the mainnet that is inside the founders reward period.
+    let block =
+        Arc::<Block>::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_415000_BYTES[..])
+            .expect("block should deserialize");
+    let mut block = Arc::try_unwrap(block).expect("block should unwrap");
+
+    // Remove coinbase transaction
+    block.transactions.remove(0);
+
+    // Validate the block
+    let result = check::subsidy_is_correct(network, &block).unwrap_err();
+    let expected = BlockError::Transaction(TransactionError::Subsidy(SubsidyError::NoCoinbase));
+    assert_eq!(expected, result);
+
+    Ok(())
+}
+
+#[test]
+fn founders_reward_validation_failure() -> Result<(), Report> {
+    use crate::error::*;
+    use zebra_chain::transaction::Transaction;
+
+    let network = Network::Mainnet;
+
+    // Get a header from a block in the mainnet that is inside the founders reward period.
+    let header =
+        block::Header::zcash_deserialize(&zebra_test::vectors::HEADER_MAINNET_415000_BYTES[..])
+            .unwrap();
+
+    // From the same block get the coinbase transaction
+    let block =
+        Arc::<Block>::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_415000_BYTES[..])
+            .expect("block should deserialize");
+    let block = Arc::try_unwrap(block).expect("block should unwrap");
+    let coinbase_transaction = block.transactions.get(0).unwrap().clone();
+    let coinbase_transaction = Arc::try_unwrap(coinbase_transaction).unwrap_err();
```

Comment:
There is probably a better way to do this ...

Reply:
a lot of these look unnecessary, I think you can just completely remove lines 217 and 219 without any errors, and probably a few of the `clone` calls as well. This seems to work and do the same thing your code does:

```rust
    // Build the new transaction with modified coinbase outputs
    let tx = block
        .transactions
        .get(0)
        .map(|transaction| Transaction::V3 {
            inputs: transaction.inputs().to_vec(),
            outputs: vec![transaction.outputs()[0].clone()],
            lock_time: transaction.lock_time(),
            expiry_height: transaction.expiry_height().unwrap(),
            joinsplit_data: None,
        })
        .unwrap();
```

Reply:
You are the best, thanks :+1: